include/npcomp/RefBackend/Runtime/UserAPI.h

```diff
@@ -25,13 +25,40 @@
 #include <atomic>
 #include <cstdlib>
 #include <string>
+#include <iostream>
 
 namespace refbackrt {
 
 struct RtValue;
 
+// Generic base class for representing an object that should be reference
+// counted.
+class RefTarget {
+public:
+  std::atomic<int> refCount{0};
+  virtual ~RefTarget() = 0;
+};
+
+// Generic Base class that holds any static operations for RefCounted objects
+class RefBase {
+public:
+  static void incref(RefTarget *ptr) {
+    if (!ptr)
+      return;
+    ptr->refCount += 1;
+  }
+  static void decref(RefTarget *ptr) {
+    if (!ptr)
+      return;
+    if (ptr->refCount.fetch_sub(1) == 1) {
+      ptr->~RefTarget();
+      std::free(static_cast<void *>(ptr));
+    }
+  }
+};
```

Comment:
@silvasean: Here's a first pass at a `RefTarget` level of abstraction. However, there's a bug here in destructor chaining for `RefTarget`'s needing access to the Derived class.

It looks like there's an [llvm::IntrusiveRefCountPtr](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ADT/IntrusiveRefCntPtr.h) type that might work well, although the base class requires a template type for proper destructor chaining:

https://github.com/llvm/llvm-project/blob/7a64cc4a764eb30499d8254d071e2e315a46f3dd/llvm/include/llvm/ADT/IntrusiveRefCntPtr.h#L71-L98

However, there's no `IntrusiveRefCountPtrTarget` abstraction, which still leaves us stuck with a `void*` in the payloadUnion, or a series of ptrs per data structure (`Tensor* asTensor; List* asList;`, etc.)

Digging into `c10::intrusive_ptr` and their abstraction layers to figure out how they solved.

Reply:
Good catch! That's subtle.

`intrusive_ptr_target` "kind of" destructs via the [release_resources](https://github.com/pytorch/pytorch/blob/25a9f45a5adfa0e717e95ca7cf94086950f3b4d1/c10/util/intrusive_ptr.h#L143) virtual function. It's not a full destruction technically but it does free any resources such that a regular free of the memory block is enough (all transitively owned resources are already freed by release_resources). We could do something analogous just to keep it simple and braindead.

This is low-level runtime code and can't depend on heavy libSupport-based abstractions like llvm::IntrusiveRefCountPtr for code size reasons. (also llvm::IntrusiveRefCountPtr doesn't have the virtual function analogous to `release_resources` so it wouldn't help here.).